### PR TITLE
Factorize std model

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description
Anki code contains the comment
> // AnkiDroid doesn't have stdmodels class or model name localization, this could be much cleaner if implemented

One part was fixed in earlier commit: name localization
This one fixes the absence of stdmodels

stdmodels do more here than in Anki, because java is more verbose, because of the way it deals with optional parameters (named parameters in Python). So the class StdModel allows to create new std models, add them to the collection, and contains static StdModel variable for each standard model, and a list of standard model.

This allow to loop over the list of standard models instead of having to add them manually, and this allow to access the resources at a single point. 

## Approach
Creating a class, whose constructor takes as parameter the default name (as a resource, not as a String, so it can be localized), and a function constructing a new model (once again, so that the actual model can be localized)
This class contains as static variables all standard models and a list of standard models. Those are then used everywhere they are needed in the code

This also allows to remove the magical constant 4, replacing it by the size of the array.

## How Has This Been Tested?

In the VM, I opened the note type manager. I added a new standard model and saw it works. I also cloned a model and saw that the correct model was cloned. In both case, I did check the list of fields to ensure the correct things was cloned.

## Learning (optional, can help others)
It may be better doing code cleaning before doing the actual change I care about.
Because having this would have simplified #5572 a lot and slightly #5493. 

I don't know which PR will be merged first, but in any case, i'll have work to do to ensure that the two other PR can be merged.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
In this case, I'll disagree with Codacy. So I'll explain why.
RuntimeException: this is the way JSONException as dealt with everywhere in the code. So I believe it should be accepted here too (at least until the JSON PR is merged)
The method name _new is actually kind of standard if we try to follow name used in Python.

The static fields uses methods defined in this class. I do believe that it's more beautiful if they are defined after the method they use.

**Comment by mikehardy** [comment]
This would fail a local test based on the CI failure

**Comment by Arthur-Milchior** [comment]
I assure you. If I check out this branch, with a VM started (API22) and do `./gradlew clean jacocoUnitTestReport` I see no error message related to a test. 


However, I get 
>[ant:jacocoReport] Classes in bundle 'AnkiDroid' do no match with execution data. For report generation the same class files must be used as at runtime.
[ant:jacocoReport] Execution data for class com/wildplot/android/rendering/graphics/wrapper/ColorWrap does not match.

which I don't understand.

**Comment by mikehardy** [comment]
> I assure you. If I check out this branch, with a VM started (API22) and do `./gradlew clean jacocoUnitTestReport` I see no error message related to a test.

If you are doing jacoco**Unit**TestReport, it's just the "unit" tests. Not the "instrumented" tests that actually engage the emulator. You can pass jacocoUnitTestReport all day with no emulator running. In fact, if you try jacocoTestReport (which attempts to run both) with no emulator up, the instrumented tests don't run (no emulator...) so they pass too. We have 2 main categories of tests, and that's how you differentiate.

That said, tests are passing now


> 
> However, I get
> 
> > [ant:jacocoReport] Classes in bundle 'AnkiDroid' do no match with execution data. For report generation the same class files must be used as at runtime.
> > [ant:jacocoReport] Execution data for class com/wildplot/android/rendering/graphics/wrapper/ColorWrap does not match.
> 
> which I don't understand.

That's an artifact of "parameterized" testing being done on the stats object and is luckily inconsequential because fixing it seems intractable. Only affects test coverage data (something interesting but not vital), and only on that class

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java  3
         

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639&bid=14883889)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    public String getDefaultName() {
+        return AnkiDroidApp.getAppResources().getString(defaultName);
+    }
+
+
+    /// create the standard models
+
+    public static final StdModels basicModel = new StdModels(
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    interface CreateStdModels {
+        JSONObject create(Models mm, String name);
+    }
+
+    public StdModels(CreateStdModels fun, @StringRes int defaultName) {
+        this.fun = fun;
+        this.defaultName = defaultName;
+    }
+
+    private JSONObject _new(Models mm) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [The instance method name '_new' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            JSONObject t = m.getJSONArray("tmpls").getJSONObject(1);
+            t.put("qfmt", "{{#" + av +"}}" + t.get("qfmt") + "{{/" + av +"}}");
+        } catch (JSONException e) {
+            throw new RuntimeException(e);
+        }
+        return m;
+    },
+        R.string.forward_optional_reverse_model_name);
+
+    public static final StdModels clozeModel = new StdModels
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            t.put("afmt", fmt + "<br>\n{{" + fieldExtraName + "}}");
+            mm.addTemplateInNewModel(m, t);
+        } catch (JSONException e) {
+            throw new RuntimeException(e);
+        }
+        return m;
+    },
+        R.string.cloze_model_name);
+
+    public static StdModels[] stdModels =
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            mm.addFieldInNewModel(m, fm);
+            String cardTypeClozeName = AnkiDroidApp.getAppResources().getString(R.string.card_cloze_name);
+            JSONObject t = mm.newTemplate(cardTypeClozeName);
+            String fmt = "{{cloze:" + txt + "}}";
+            m.put("css", m.getString("css") + ".cloze {" + "font-weight: bold;" + "color: blue;" + "}");
+            t.put("qfmt", fmt);
+            t.put("afmt", fmt + "<br>\n{{" + fieldExtraName + "}}");
+            mm.addTemplateInNewModel(m, t);
+        } catch (JSONException e) {
+            throw new RuntimeException(e);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                    t.put("afmt", "{{FrontSide}}\n\n<hr id=answer>\n\n{{" + backName + "}}");
+                } catch (JSONException e) {
+                    throw new RuntimeException(e);
+                }
+                mm.addTemplateInNewModel(m, t);
+                return m;
+            },
+            R.string.basic_model_name);
+
+    public static final StdModels forwardReverseModel = new StdModels
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        this.fun = fun;
+        this.defaultName = defaultName;
+    }
+
+    private JSONObject _new(Models mm) {
+        String name = getDefaultName();
+        return _new(mm, name);
+    }
+
+    private JSONObject _new(Models mm, String name) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [The instance method name '_new' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

File: AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            t.put("afmt", "{{FrontSide}}\n\n<hr id=answer>\n\n{{"+frontName+"}}");
+            mm.addTemplateInNewModel(m, t);
+        } catch (JSONException e) {
+            throw new RuntimeException(e);
+        }
+        return m;
+    },
+        R.string.forward_reverse_model_name);
+
+    public static final StdModels forwardOptionalReverseModel = new StdModels
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639)

**Comment by Arthur-Milchior** [comment]
I don't know whether I'm expected to answer to all/some comments which were generated by Codacy. 
Because most of them does not makes sens. For examples: 
* The instance method name '_new' doesn't match '[a-z][a-zA-Z0-9]*' which I already mentioned a lot. This rule is false when we follow Anki's name
* Avoid throwing raw exception types. This is actually what is done in all of ankidroid. There seems to be nothing else to do. Unless you merge another PR I made about json
* Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes. I can reorder those fields, but I believe that staying kind of close to anki's code should be best here.

**Comment by mikehardy** [comment]
@Arthur-Milchior no worries - libanki - where you have primarily been working - is a special case because it follows all sorts of Python conventions vs Java, and it's hard to tune codacy to handle that. Your specific notes are all correct to ignore for libanki

**Comment by mikehardy** [comment]
@Arthur-Milchior I think this is ready to go in general as well, except you had another PR that unavoidably conflicted with it in #5493 - I just merged that one, which I think will clear this one up to go if you want to patch it up

**Comment by Arthur-Milchior** [comment]
Rebased onto master

**Comment by Arthur-Milchior** [comment]
Rebased

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 8
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/libanki/StdModels.java  1
         

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4389639&bid=14883889)

**Pull request merged**